### PR TITLE
[ANOMALY_CLASSIFICATION] Anomaly score normalization

### PR DESCRIPTION
**Pull request by djdameln**
The PR adds anomaly score normalization to the Anomaly Classification task. This ensures that the predicted confidence scores are always between 0 and 1. It also improves the quality of the saliency maps, especially for normal images.

- Use `AnomalyScoreNormalizationCallback` from anomalib to normalize predictions.
- Store anomaly score distribution of training set in output model (needed for OpenVINO post-processing).
- In OpenVINO inference, pass meta data to post processor to enable normalization.
- Improve test case to check if OpenVINO task produces the same output as the standard task.
- Increase anomalib version to 0.2.4

BMM Build: [22047](https://ci.iotg.sclab.intel.com/job/IMPT/job/IMPTOps/22047/) 🕐

### CODE REVIEW

File: external/anomaly/anomaly_classification/task.py [hunk 1]
```diff
@@ -154,9 +155,15 @@ def save_model(self, output_model: ModelEntity):
         output_model.set_data("weights.pth", buffer.getvalue())
         output_model.set_data("label_schema.json", label_schema_to_bytes(self.task_environment.label_schema))
         # store computed threshold
-        output_model.set_data("threshold", bytes(struct.pack("f", self.model.threshold.item())))
-
-        f1_score = self.model.image_metrics.OptimalF1.compute().item()
+        output_model.set_data("threshold", bytes(struct.pack("f", self.model.image_threshold.value.item())))
+        output_model.set_data("pixel_threshold", bytes(struct.pack("f", self.model.pixel_threshold.value.item())))
```
Comment (samet-akcay) [inline external/anomaly/anomaly_classification/task.py]: Similar to what you've done in Lines 160-164, would it be possible to follow the same convention here to convert Lines 158-159 to bytes. L160-164 looks simpler to read.

Comment (samet-akcay) [reply external/anomaly/anomaly_classification/task.py]: Something like

```
output_model.set_data("threshold", self.model.image_threshold.value.cpu().numpy().tobytes())
output_model.set_data("pixel_threshold", self.model.pixel_threshold.value.cpu().numpy().tobytes())
```

File: external/anomaly/anomaly_classification/task.py [hunk 2]
```diff
@@ -227,7 +235,7 @@ def export(self, export_type: ExportType, output_model: ModelEntity):
         with open(xml_file, "rb") as file:
             output_model.set_data("openvino.xml", file.read())
         output_model.set_data("label_schema.json", label_schema_to_bytes(self.task_environment.label_schema))
-        output_model.set_data("threshold", bytes(struct.pack("f", self.model.threshold.item())))
+        output_model.set_data("threshold", bytes(struct.pack("f", self.model.image_threshold.value.item())))
```
Comment (samet-akcay) [inline external/anomaly/anomaly_classification/task.py]: Similar to above. Is it possible to do the following?

```
output_model.set_data("threshold", self.model.image_threshold.value.cpu().numpy().tobytes())

```

**Review by samet-akcay — COMMENTED** [review]
Thanks @djdameln ! I've done my first round of review

**Comment by samet-akcay** [comment]
run ote_sdk tests

**Review by ashwinvaidya17 — DISMISSED** [review]

**Comment by Ilya-Krylov** [comment]
These things should be probably updated as well

https://github.com/openvinotoolkit/training_extensions/pull/782#discussion_r773719865 
https://github.com/openvinotoolkit/training_extensions/pull/782#discussion_r773719439

**Comment by samet-akcay** [comment]
run ote_sdk tests

File: external/anomaly/anomaly_classification/task.py (hunk 2 above)
Comment (Ilya-Krylov) [reply external/anomaly/anomaly_classification/task.py]: can it be "confidence_threshold" as in case of detection. Just would like to have these things aligned

Comment (samet-akcay) [reply external/anomaly/anomaly_classification/task.py]: hmm, this would be an idea to distinguish the term threshold from image_threshold and pixel_threshold. @djdameln , thoughts?

Comment (djdameln) [reply external/anomaly/anomaly_classification/task.py]: This threshold value is applied to the (standardized) anomaly scores, so technically it is not a confidence threshold. The anomaly scores are converted to confidence values after the threshold is applied. That's why I prefer to keep using `threshold`. I could change it to `image_threshold` to make the distinction between image-level and pixel-level threshold and better match the naming in anomalib.

**Comment by Ilya-Krylov** [comment]
fix tests please

**Comment by djdameln** [comment]
run ote_sdk tests

**Comment by djdameln** [comment]
run ote_sdk tests

**Comment by djdameln** [comment]
run ote_sdk tests

File: ote_sdk/ote_sdk/usecases/exportable_code/prediction_to_annotation_converter.py [hunk 3]
```diff
@@ -272,10 +272,9 @@ def __init__(self, label_schema: LabelSchemaEntity):
         ]
 
     def convert_to_annotation(
-        self, predictions: np.ndarray, metadata: Dict[str, Any]
+        self, pred_score: float, metadata: Dict[str, Any]
```
Comment (Ilya-Krylov) [inline ote_sdk/ote_sdk/usecases/exportable_code/prediction_to_annotation_converter.py:275]: Please follow the interface

File: ote_cli/ote_cli/utils/io.py [hunk 4]
```diff
@@ -62,7 +62,7 @@ def read_model(model_configuration, path, train_dataset):
             "openvino.bin": ModelAdapter(read_binary(path[:-4] + ".bin")),
         }
 
-        for confidence_threshold in ["confidence_threshold", "threshold"]:
+        for confidence_threshold in ["confidence_threshold", "image_threshold"]:
```
Comment (Ilya-Krylov) [inline ote_cli/ote_cli/utils/io.py:65]: You can add other metadata keys (image_mean, image_std, pixel_mean, ..) so they can be loaded

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests

**Comment by Ilya-Krylov** [comment]
close?

**Comment by ashwinvaidya17** [comment]
> close?

@Ilya-Krylov This contains changes related to another normalization method. Till the time it does not delete the branch, closing should be okay. We might need these changes in case the this method is required to be added later.
